src/app/util/remote-data.spec.ts

```diff
@@ -0,0 +1,161 @@
+      const data = "Data";
+      const remoteData = success(data);
+      expect(remoteData.kind).toEqual(RemoteDataKind.Success);
+      expect(remoteData.value).toEqual(data);
+      expect(remoteData.isLoading()).toBeFalsy();
+      expect(remoteData.hasData()).toBeTruthy();
+    });
+  });
+  describe("loading", () => {
+    it("should create loading when given no argument", () => {
+      const remoteData = loading();
+      expect(remoteData.kind).toEqual(RemoteDataKind.Loading);
+      expect(remoteData.isLoading()).toBeTruthy();
+      expect(remoteData.hasData()).toBeFalsy();
+    });
+    it("should create loading when given notAsked", () => {
+      const remoteData = loading(notAsked());
+      expect(remoteData.kind).toEqual(RemoteDataKind.Loading);
+      expect(remoteData.isLoading()).toBeTruthy();
+      expect(remoteData.hasData()).toBeFalsy();
+    });
+    it("should create reloading when given success", () => {
+      const data = "Data";
+      const remoteData = loading(success(data));
+      if (remoteData.kind === RemoteDataKind.Reloading) {
+        expect(remoteData.isLoading()).toBeTruthy();
+        expect(remoteData.value).toEqual(data);
+        expect(remoteData.hasData()).toBeTruthy();
+      } else {
+        expect(remoteData.kind).toEqual(RemoteDataKind.Reloading);
+      }
+    });
+    it("should create reloading when given error with data", () => {
+      const data = "Data";
+      const remoteData = loading(error("Error", success(data)));
+      if (remoteData.kind === RemoteDataKind.Reloading) {
+        expect(remoteData.isLoading()).toBeTruthy();
+        expect(remoteData.value).toEqual(data);
+        expect(remoteData.hasData()).toBeTruthy();
+      } else {
+        expect(remoteData.kind).toEqual(RemoteDataKind.Reloading);
+      }
+    });
+    it("should create loading when given error with no data", () => {
+      const data = "Data";
+      const remoteData = loading(error("Error"));
+      if (remoteData.kind === RemoteDataKind.Loading) {
+        expect(remoteData.isLoading()).toBeTruthy();
+        expect(remoteData.hasData()).toBeFalsy();
+      } else {
+        expect(remoteData.kind).toEqual(RemoteDataKind.Loading);
+      }
+    });
+    it("should create loading when given loading", () => {
+      const remoteData = loading(loading());
+      expect(remoteData.kind).toEqual(RemoteDataKind.Loading);
+      expect(remoteData.isLoading()).toBeTruthy();
+      expect(remoteData.hasData()).toBeFalsy();
+    });
+    it("should create reloading when given reloading", () => {
+      const data = "Data";
+      const remoteData = loading(loading(success(data)));
+      if (remoteData.kind === RemoteDataKind.Reloading) {
+        expect(remoteData.value).toEqual(data);
+        expect(remoteData.isLoading()).toBeTruthy();
+        expect(remoteData.hasData()).toBeTruthy();
+      } else {
+        expect(remoteData.kind).toEqual(RemoteDataKind.Reloading);
```

Comment:
In the test, wouldn't these perform the same every time? How would each branch of if else statement get hit?

Reply:
the `expect` on the `else` branch is just to make the test fail in case it hit that case. I originally put `fail()` there, but I changed it  to this version. I don't know which one easier to understand. Probably `fail()` since this version raised a question already?